File: python/tools/latn_iast_transliteration_verification/latn_iast_to_deva_test.py
from __future__ import annotations

from indic_script_converter.latn_iast_to_deva import (
    DevanagariRomanizationProfile,
    IastToDevanagariDigitPolicy,
    IastToDevanagariOmPolicy,
    IastToDevanagariOptions,
    IastToDevanagariPunctuationPolicy,
    to_devanagari_from_iast,
)
from indic_script_converter.tools.example_latn_iast import (
    transliteration_smoke_samples,
    vedic_round_trip_cases,
)


def main() -> None:
    print('----------------------------------------------------------------')
    print(' 1. DEVANAGARI TRANSLITERATOR SAMPLES')
    print('----------------------------------------------------------------')

    default_options = IastToDevanagariOptions(
        punctuation_policy=IastToDevanagariPunctuationPolicy.INDIC_DANDA,
        digit_policy=IastToDevanagariDigitPolicy.CONVERT_TO_SCRIPT,
    )

    for source in transliteration_smoke_samples:
        result = to_devanagari_from_iast(source, default_options)
        print(f'"{source}" -> "{result}"')

    _run_option_tests()
    _run_profile_tests()
    _run_vedic_round_trip_tests()


def _run_option_tests() -> None:
    print('\n[Devanagari Option: Digits Policy (convertToScript)]')
    digit_options = IastToDevanagariOptions(
        digit_policy=IastToDevanagariDigitPolicy.CONVERT_TO_SCRIPT,
    )
    digit_sample = '12345'
    print(f'  "{digit_sample}" -> "{to_devanagari_from_iast(digit_sample, options=digit_options)}"')

    print('\n[Devanagari Option: Danda Policy (indicDanda)]')
    danda_options = IastToDevanagariOptions(
        punctuation_policy=IastToDevanagariPunctuationPolicy.INDIC_DANDA,
    )
    punctuation_sample = 'End. Double end..'
    print(
        f'  "{punctuation_sample}" -> '
        f'"{to_devanagari_from_iast(punctuation_sample, options=danda_options)}"'
    )

    print('\n[Devanagari Option: OM Policy (useOmSign)]')
    om_options = IastToDevanagariOptions(
        om_policy=IastToDevanagariOmPolicy.USE_OM_SIGN,
    )
    om_sample = 'oṃ namaḥ śivāya'
    print(f'  "{om_sample}" -> "{to_devanagari_from_iast(om_sample, options=om_options)}"')


def _run_profile_tests() -> None:
    print('\n[Devanagari Profiles: strictIast / iso15919Core / extendedIndic]')
    strict = IastToDevanagariOptions(
        profile=DevanagariRomanizationProfile.STRICT_IAST,
    )
    iso = IastToDevanagariOptions(
        profile=DevanagariRomanizationProfile.ISO_15919_CORE,
    )
    extended = IastToDevanagariOptions(
        profile=DevanagariRomanizationProfile.EXTENDED_INDIC,
    )

    for source in ['ṛaka', 'ṛhaṇa', 'ḷa', 'laṛkā', 'xaṇḍa']:
        print(
            f'  "{source}" (strictIast)    -> "{to_devanagari_from_iast(source, options=strict)}"'
        )
        print(f'  "{source}" (iso15919Core) -> "{to_devanagari_from_iast(source, options=iso)}"')
        print(
            f'  "{source}" (extendedIndic)-> "{to_devanagari_from_iast(source, options=extended)}"'
        )


def _run_vedic_round_trip_tests() -> None:
    print('\n================================================================')
    print(' VEDIC FIXTURES: IAST → Devanagari vs expected Devanagari')
    print('================================================================')

    options = IastToDevanagariOptions(
        punctuation_policy=IastToDevanagariPunctuationPolicy.INDIC_DANDA,
        digit_policy=IastToDevanagariDigitPolicy.CONVERT_TO_SCRIPT,
    )

    passed = 0
    failed = 0

    for iast, expected, label in vedic_round_trip_cases:
        actual = to_devanagari_from_iast(iast, options=options)
        if actual == expected:
            passed += 1
            print(f'  ✓ {label}')
            continue

        failed += 1
        print(f'  ✗ {label}')
        print(f'    DIFF: {_rune_diff(actual, expected)}')
        print(f'    GOT: {actual}')
        print(f'    EXP: {expected}')

    print(
        f'\n  Result: {passed} passed, {failed} failed out of '
        f'{len(vedic_round_trip_cases)} fixtures.'
    )

    if failed > 0:
        raise ValueError(f'{failed} Vedic fixture(s) failed.')


def _rune_diff(actual: str, expected: str) -> str:
    actual_runes = [ord(ch) for ch in actual]
    expected_runes = [ord(ch) for ch in expected]
    length = max(len(actual_runes), len(expected_runes))
    output = []

    for index in range(length):
        actual_rune = actual_runes[index] if index < len(actual_runes) else None
        expected_rune = expected_runes[index] if index < len(expected_runes) else None
        if actual_rune == expected_rune:
            continue

        output.append(f'[{_format_rune(actual_rune)}|{_format_rune(expected_rune)}@{index}]')

    return ' '.join(output)


def _format_rune(rune: int | None) -> str:
    if rune is None:
        return 'END'
    return f'U+{hex(rune)[2:].upper().zfill(4)}'


if __name__ == '__main__':
    main()

File: indic-script-converter-js/tools/latn_iast_transliteration_verification/latn-iast-to-deva-test.js
import {
    DevanagariRomanizationProfile,
    IastToDevanagariDigitPolicy,
    IastToDevanagariOptions,
    IastToDevanagariPunctuationPolicy,
    IastToDevanagariOmPolicy,
    toDevanagariFromIast,
    transliterationSmokeSamples,
    vedicRoundTripCases,
} from '../../src/index.js';

export function main() {
    console.log('----------------------------------------------------------------');
    console.log(' 1. DEVANAGARI TRANSLITERATOR SAMPLES');
    console.log('----------------------------------------------------------------');

    const defaultOptions = new IastToDevanagariOptions({
        punctuationPolicy: IastToDevanagariPunctuationPolicy.INDIC_DANDA,
        digitPolicy: IastToDevanagariDigitPolicy.CONVERT_TO_SCRIPT,
    });

    for (const source of transliterationSmokeSamples) {
        console.log(`"${source}" -> "${toDevanagariFromIast(source, defaultOptions)}"`);
    }

    runOptionTests();
    runProfileTests();
    runVedicRoundTripTests();
}

function runOptionTests() {
    console.log('\n[Devanagari Option: Digits Policy (convertToScript)]');
    const digitOptions = new IastToDevanagariOptions({
        digitPolicy: IastToDevanagariDigitPolicy.CONVERT_TO_SCRIPT,
    });
    const digitSample = '12345';
    console.log(`  "${digitSample}" -> "${toDevanagariFromIast(digitSample, digitOptions)}"`);

    console.log('\n[Devanagari Option: Danda Policy (indicDanda)]');
    const dandaOptions = new IastToDevanagariOptions({
        punctuationPolicy: IastToDevanagariPunctuationPolicy.INDIC_DANDA,
    });
    const punctuationSample = 'End. Double end..';
    console.log(
        `  "${punctuationSample}" -> "${toDevanagariFromIast(punctuationSample, dandaOptions)}"`,
    );

    console.log('\n[Devanagari Option: OM Policy (useOmSign)]');
    const omOptions = new IastToDevanagariOptions({
        omPolicy: IastToDevanagariOmPolicy.USE_OM_SIGN,
    });
    const omSample = 'oṃ namaḥ śivāya';
    console.log(`  "${omSample}" -> "${toDevanagariFromIast(omSample, omOptions)}"`);
}

function runProfileTests() {
    console.log('\n[Devanagari Profiles: strictIast / iso15919Core / extendedIndic]');
    const strict = new IastToDevanagariOptions({
        profile: DevanagariRomanizationProfile.STRICT_IAST,
    });
    const iso = new IastToDevanagariOptions({
        profile: DevanagariRomanizationProfile.ISO_15919_CORE,
    });
    const extended = new IastToDevanagariOptions({
        profile: DevanagariRomanizationProfile.EXTENDED_INDIC,
    });

    for (const source of ['ṛaka', 'ṛhaṇa', 'ḷa', 'laṛkā', 'xaṇḍa']) {
        console.log(`  "${source}" (strictIast)    -> "${toDevanagariFromIast(source, strict)}"`);
        console.log(`  "${source}" (iso15919Core) -> "${toDevanagariFromIast(source, iso)}"`);
        console.log(`  "${source}" (extendedIndic)-> "${toDevanagariFromIast(source, extended)}"`);
    }
}

function runVedicRoundTripTests() {
    console.log('\n================================================================');
    console.log(' VEDIC FIXTURES: IAST → Devanagari vs expected Devanagari');
    console.log('================================================================');

    const options = new IastToDevanagariOptions({
        punctuationPolicy: IastToDevanagariPunctuationPolicy.INDIC_DANDA,
        digitPolicy: IastToDevanagariDigitPolicy.CONVERT_TO_SCRIPT,
    });

    let passed = 0;
    let failed = 0;

    for (const [iast, expected, label] of vedicRoundTripCases) {
        const actual = toDevanagariFromIast(iast, options);
        if (actual === expected) {
            passed += 1;
            console.log(`  ✓ ${label}`);
            continue;
        }

        failed += 1;
        console.log(`  ✗ ${label}`);
        console.log(`    DIFF: ${runeDiff(actual, expected)}`);
        console.log(`    GOT: ${actual}`);
        console.log(`    EXP: ${expected}`);
    }

    console.log(
        `\n  Result: ${passed} passed, ${failed} failed out of ${vedicRoundTripCases.length} fixtures.`,
    );

    if (failed > 0) {
        throw new Error(`${failed} Vedic fixture(s) failed.`);
    }
}

function runeDiff(actual, expected) {
    const actualRunes = [...actual].map((c) => c.codePointAt(0));
    const expectedRunes = [...expected].map((c) => c.codePointAt(0));
    const length = Math.max(actualRunes.length, expectedRunes.length);
    const output = [];

    for (let index = 0; index < length; index++) {
        const actualRune = index < actualRunes.length ? actualRunes[index] : null;
        const expectedRune = index < expectedRunes.length ? expectedRunes[index] : null;
        if (actualRune === expectedRune) continue;

        output.push(`[${formatRune(actualRune)}|${formatRune(expectedRune)}@${index}]`);
    }

    return output.join(' ');
}

function formatRune(rune) {
    if (rune === null) return 'END';
    return `U+${rune.toString(16).toUpperCase().padStart(4, '0')}`;
}

if (import.meta.url === `file://${process.argv[1]}`) main();
